Autofill: Implement mapping for appPackages for autofill provider service

karlenDimla opened the pull request:
Task/Issue URL: https://app.asana.com/0/1201462763415876/1209125943914364/f

### Description
See attached task description

### Steps to test this PR
https://app.asana.com/0/1201462763415876/1209196674121696/f

karlenDimla commented:
Seems like the lint failures and unit test failures were on your changes @cmonfortep  so I wouldn’t touch them.

karlenDimla commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/AppCredentialProvider.kt:
```diff
@@ -0,0 +1,55 @@
+    private val autofillStore: AutofillStore,
+) : AppCredentialProvider {
+    override suspend fun getCredentials(appPackage: String): List<LoginCredentials> = withContext(dispatcherProvider.io()) {
+        Timber.d("Autofill-mapping: Getting credentials for $appPackage")
```
Ah good catch. Not actually sure why I used timber here in the first place. Will fix

cmonfortep replied:
is not autofill module using `Timber`? I believe logcat is the one to be removed as it's added as new dependency.

karlenDimla replied:
Will update to Timber as @cmonfortep now changed the scope used for the service.

cmonfortep commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/RemoteDomainTargetAppDataDownloader.kt:
```diff
@@ -0,0 +1,99 @@
+) : MainProcessLifecycleObserver {
+    override fun onCreate(owner: LifecycleOwner) {
+        super.onCreate(owner)
+        // TODO: Add check for killswitch
```
is this a recommendation or a todo we should do? I can add it later but leave me a task in Asana as pending.

karlenDimla replied:
Synced about this. @cmonfortep  will do this once he add the killswitch

cmonfortep commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/RemoteDomainTargetAppService.kt:
```diff
@@ -0,0 +1,43 @@
+@ContributesServiceApi(AppScope::class)
+interface RemoteDomainTargetAppService {
+    // TODO: replace with real URL
+    @GET("https://www.jsonblob.com/api/1329813836072673280")
```
We should update this to the real endpoint.

karlenDimla replied:
WIP

karlenDimla replied:
Done

cmonfortep reviewed (commented):
Thanks @karlenDimla for the work here.
Left some comments. We need to rebase so the diff gets updated

karlenDimla commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/PackageManagerExtension.kt:
```diff
@@ -0,0 +1,87 @@
+        // Get the signing certificates
+        val signatures = packageInfo.signingInfo.apkContentsSigners
+
+        signatures.firstOrNull()?.sha256()
```
Ah good point. Im was not entirely sure how this API performs, but after reading the API, you are right that possibly multiple certificates can be used to sign one app. Updated

karlenDimla commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/AssetLinksLoader.kt:
```diff
@@ -0,0 +1,60 @@
+
+    companion object {
+        private const val ASSET_LINKS_PATH = "/.well-known/assetlinks.json"
+        private const val LOGIN_CREDENTIALS_RELATION = "delegate_permission/common.get_login_creds"
```
Fixed

karlenDimla commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/AppFingerprintProvider.kt:
```diff
@@ -0,0 +1,36 @@
+    private val context: Context,
+) : AppFingerprintProvider {
+    override fun getSHA256HexadecimalFingerprint(packageName: String): String? =
+        context.packageManager.getSHA256HexadecimalFingerprintCompat(packageName, appBuildConfig)
```
The extension function `context.packageManager.getSHA256HexadecimalFingerprintCompat` in `PackageManagerExtension` have a runcatching

karlenDimla commented:
Addressed all comments (pending final endpoint). if you don’t mind manually running this https://app.asana.com/0/1201462763415876/1209196674121696/f would be helpful. Thanks! @cmonfortep @CDRussell

cmonfortep commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/AppCredentialProvider.kt:
```diff
@@ -0,0 +1,56 @@
+interface AppCredentialProvider {
+    /**
+     * Provide a list of unique credentials that can be associated with the given [appPackage]
+     * You DO NOT need to set any dispatcher to call this suspend function
```
NIT: `You DO NOT need to...` can be removed imho, it should be safe to call any suspend method from any dispatcher (and we do that here)

cmonfortep commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/AppToDomainMapper.kt:
```diff
@@ -0,0 +1,126 @@
+interface AppToDomainMapper {
+    /**
+     * Returns a list of domains whose credentials can be associated to the [appPackage]
+     * You DO NOT need to set any dispatcher to call this suspend function
```
ditto

cmonfortep commented on autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/mapper/RemoteDomainTargetAppService.kt:
```diff
@@ -0,0 +1,42 @@
+
+@ContributesServiceApi(AppScope::class)
+interface RemoteDomainTargetAppService {
+    @GET("https://ddg-staticcdn.s3.amazonaws.com/android/domain-app-mapping.json")
```
https://app.asana.com/0/0/1209234685955829/1209248134165550/f

cmonfortep reviewed (commented):
Code review final pass, all good from my side 👍 

Planning to test this today afternoon, and report back

karlenDimla commented:
@cmonfortep Awesome! Feel free to merge then when ready. Thanks

cmonfortep commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5508?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5508** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5508?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5508?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#5561** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5561?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->
